reader-osm/src/main/java/com/graphhopper/reader/osm/OSMReader.java

```diff
@@ -720,12 +694,15 @@ EdgeIteratorState addEdge(int fromIndex, int toIndex, PointList pointList, IntsR
 
         EdgeIteratorState iter = graph.edge(fromIndex, toIndex).setDistance(towerNodeDistance).setFlags(flags);
 
-        if (nodes > 2) {
+        if (pointList.size() > 2) {
             if (doSimplify)
-                simplifyAlgo.simplify(pillarNodes);
+                simplifyAlgo.simplifyExcludeFirstLast(pointList);
 
-            iter.setWayGeometry(pillarNodes);
+            // If the entire way is just the first and last point, do not waste space storing an empty way geometry
+            if (!pointList.isEmpty())
+                iter.setWayGeometry(pointList);
```

Comment:
at first I was creating a shallow copy excluding first and last points, but then I tried this approach of removing the first and last point as part of simplification and it was consistently faster than the shallow copy approach on a medium size ~1GB extract (US northeast)

Reply:
alright, looks like that wasn't safe because it needs to remove first and last point whether you simplify or not.